### ISSUE TITLE
API: identical/equality (`__eq__`) to consider NaN coordinates as equal

### ISSUE THREAD

**Issue by jorisvandenbossche** [shapely/shapely#1775]
Opening a dedicated issue about this to highlight it, instead of only discussing it as part of the PR https://github.com/shapely/shapely/pull/1760. That PR is fixing a regression in `__eq__` in shapely 2.0 where it now accidentally ignores the z dimension: https://github.com/shapely/shapely/issues/1732 (because of the use of `equals_exact`, which apparently only considers the x and y dimensions). 

While fixing this regression, this limitation of EqualsExact was brought up with GEOS, which led to the addition of a `GEOSEqualsIdentical` (https://github.com/libgeos/geos/pull/810). However, for this new function that checks if two geometries are identical (so not just spatially equivalent, but exact pointwise equivalence (without any tolerance) for all dimensions), also NaNs in the same location in the coordinate sequence are considered as "identical".

Currently, we consider geometries with NaNs as unequal (using shapely 2.0.1):

```python
>>> import shapely
>>> line_string_nan = shapely.LineString([(1, 2), (3, np.nan)])
>>> line_string_nan
<LINESTRING (1 2, 3 NaN)>
>>> line_string_nan == line_string_nan
False
```

While if we would use the "equals_identical" logic for `__eq__`, the above would return True (which is what I am currently doing in https://github.com/shapely/shapely/pull/1760.

With shapely 1.x, this also resulted in False, because we implemented `__eq__` as `tuple(self.coords) == tuple(other.coords)`, which ends up comparing the scalar coordinate values, and then `nan != nan`. 

---

In general, people shouldn't actually use NaNs in their coordinates (and we can do more to prevent that from happening upon construction, that's for a separate issue), and then you don't run into those questions. But _if_ (for whatever reason) you do have NaNs, one should be able to test if two geometries are identical (including their NaNs) in some way. The question then of course is whether `__eq__` needs to be this way. If we add an `equals_identical()` function, people can also use that.

My feeling it makes sense to let `__eq__` be this "identical" check, since you have `equals()` for spatial equality and `equals_exact()` if you want to be able to specify some tolerance (and which will consider NaNs as unequal). 

cc @shapely/shapely-team

**Comment by brendan-ward** [comment]
> My feeling it makes sense to let __eq__ be this "identical" check, since you have equals() for spatial equality and equals_exact() if you want to be able to specify some tolerance.

Agreed.

**Comment by tfardet** [comment]
NaN is always a pain... considering coordinates with NaN equal might make sense for some people but it will probably also go against everything other people expect as all scientific Python libraries I know use the NaN != NaN convention.

**Comment by martinfleis** [comment]
I am also not a big fan of going against the convention that `NaN != NaN`, which is what is proposed. A user will conceptualise this check as a check of coordinates and then the behaviour of shapely would be inconsistent with the check over two `.coords` via numpy.

**Comment by idanmiara** [comment]
> I am also not a big fan of going against the convention that `NaN != NaN`, which is what is proposed. A user will conceptualise this check as a check of coordinates and then the behaviour of shapely would be inconsistent with the check over two `.coords` via numpy.

There are more than two metrics to check for "equality" using Shapely. 
For me the proposed equality metric with `NaN==NaN` makes much sense: 
Say you have multiple Geometries with Z (or M, soon enough) but only some of them have the Z/M values. For me it makes sense that each two geometries are equal if they have the same Z/M values or if both don't have it (i.e Z/M are `NaN`).

**Comment by tfardet** [comment]
> Say you have multiple Geometries with Z (or M, soon enough) but only some of them have the Z/M values.

What do you mean? If geometries do not have Z/M values then we don't need to compare them so it does not matter whether NaNs are considered equal or not, does it?
On the other hand, if one has a Z/M value (even only with NaNs) while the other does not, then they should register as different IMO.

**Comment by idanmiara** [comment]
> What do you mean? If geometries do not have Z/M values then we don't need to compare them so it does not matter whether NaNs are considered equal or not, does it? On the other hand, if one has a Z/M value (even only with NaNs) while the other does not, then they should register as different IMO.

If one has Z and the other doesn't then they are different. No question here. 

The question is whether `Point(1,2,NaN)==Point(1,2,NaN)`, and I as far as I know unless we define them as equal like in the proposed solution then we'll have different results depending on the GEOS version, as with some GEOS versions you cannot differentiate between `Point(1,2,NaN)` and `Point(1,2)` as they are represented internally the same. Is this correct @jorisvandenbossche?

**Comment by idanmiara** [comment]
> From what I understand, the normal result has always been `False` in all other Shapely version and thus allowing `NaN == NaN` would actually be the breaking change.

I think that that might have been correct for Shapely 1 but can't be implemented for Shapely 2. I don't think that the value of `Point(1,2,NaN)==Point(1,2,NaN)` should very across supported GEOS versions, and I'm not sure that we can restore the old behavior that `Point(1,2,NaN)!=Point(1,2,NaN)` across different GEOS versions in Shapely 2 (unless of course we keep the dimensionality in the shapely object, which I don't think that anyone wants that)

**Comment by idanmiara** [comment]
Hi @jorisvandenbossche,
Could you please update what's the status for https://github.com/shapely/shapely/pull/1760 and what's your opinion about the discussion above ?

**Comment by jorisvandenbossche** [comment]
We chatted about this yesterday with @brendan-ward and  @martinfleis, and we would propose to indeed consider NaN == NaN if they are present in the same location in the coordinate sequence (for `Geometry.__eq__`). 

This differs from typical floating point equality which considers NaN != NaN, but in this case we are not evaluating the coordinates element-wise itself, but evaluating the equality of the full object. For that purpose, it seems useful to be able to check if two geometries are "identical" including their NaNs. 

You generally shouldn't have NaNs in your coordinates (operations will be broken), but if for some reason you have that, and want to check this, it can be useful that equality considers them the same (as "equally broken" .. ;)). For example, those two geometries will also have an equal WKB value.

This has the added benefit that `__eq__` would match with one of the `equals(_..)` functions (and so that you can always be explicit by calling the function instead), i.e. a future `equals_identical` function (if we follow that addition from GEOS).

@caspervdw @sgillies @mwtoews are you also OK with that?

Another question is: if we want to make this change, are we OK with still doing that in 2.0.2 (and consider it as one more (late) API change for 2.0), or are we fine with proposing API changes for 2.1?

**Comment by sgillies** [comment]
@jorisvandenbossche makes sense to me. Consider a geometry object `g` with one NaN coordinate. It makes no sense to have `g == g` evaluate to False. This feels like a bug fix to me and I would be fine with it in 2.0.2.

**Comment by jorisvandenbossche** [comment]
Given that we are in the meantime almost a year from the initial 2.0 release, and I hope we are closer to doing a 2.1 release, I would personally maybe lean to leaving it as a change for 2.1.
